src/capi/object.cpp

```diff
@@ -614,6 +614,40 @@ extern "C" PyObject** _PyObject_GetDictPtr(PyObject* obj) noexcept {
     }
 }
 
+// This function returns the copy of object's dict, not a writable dict.
+extern "C" PyObject* PyObject_GetDict(PyObject* obj) noexcept {
+    PyTypeObject* tp = Py_TYPE(obj);
+    if (!tp->instancesHaveHCAttrs()) {
+        if (!tp->instancesHaveDictAttrs())
+            return NULL;
+        return obj->getDict();
+    } else {
+        Box* attrwrapper = obj->getAttrWrapper();
+        // unwritable dict, this is just a copy of object's dict.
+        return attrwrapperToDict(attrwrapper);
+    }
+}
+
+extern "C" void PyObject_ClearDict(PyObject* obj) noexcept {
+    obj->clearAttrsForDealloc();
+}
+
+// This API just add or update key-value pairs. It will not remove existed items
+// if the key is not contained in the new dict.
+extern "C" void PyObject_SetDict(PyObject* obj, PyObject* dict) noexcept {
+    PyObject* d_key, *d_value;
+    Py_ssize_t i = 0;
+    while (PyDict_Next(dict, &i, &d_key, &d_value)) {
+        if (PyString_CheckExact(d_key)) {
+            Py_INCREF(d_key);
+            PyString_InternInPlace(&d_key);
+            Py_DECREF(d_key);
+        }
```

Comment:
Look like to me like this block is not necessary, `PyObject_SetAttr` should already do the interning